### PR TITLE
Implement schemas serialization

### PR DESCRIPTION
**Pull request by bugagashenkj**

### CODE REVIEW

File: lib/serialization.js [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+        node: path.node,
+        source: generate(path.node.value).code,
+        ins: getInstance(pathToNode, schema.definition),
+      };
+
+      if (exclude(options)) {
+        path.remove();
+      } else {
+        const replaceStr = replace(options);
+        if (replaceStr) {
```
Comment (nechaido) [inline lib/serialization.js]:
```suggestion
        if (replaceStr !== undefined && replaceStr !== null) {
```

File: lib/serialization.js [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+};
+
+const getInstance = (path, ins) =>
+  ins && path.length ? getInstance(path, ins[path.shift()]) : ins;
+
+const serializeSchema = (
+  schema,
+  { exclude = () => null, replace = () => null } = {}
+) => {
+  const { source, ...result } = schema;
```
Comment (nechaido) [inline lib/serialization.js]:
```suggestion
  const { source, definition, ...result } = schema;
```

File: lib/serialization.js [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+  traverse(ast, {
+    ObjectProperty(path) {
+      const pathToNode = buildPath(path).reverse();
+
+      const options = {
+        path: pathToNode,
+        key: path.node.key.name,
+        node: path.node,
+        source: generate(path.node.value).code,
+        ins: getInstance(pathToNode, schema.definition),
```
Comment (nechaido) [inline lib/serialization.js]:
```suggestion
        ins: getInstance(pathToNode, definition),
```

File: lib/serialization.js [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+      } else {
+        const replaceStr = replace(options);
+        if (replaceStr) {
+          path.node.value = parseExpression(replaceStr);
+        }
+      }
+    },
+  });
+
+  result.definition = null;
```
Comment (nechaido) [inline lib/serialization.js]:
```suggestion
```

File: test/serialization.js [hunk 5]
```diff
@@ -0,0 +1,57 @@
+'use strict';
```
Comment (lundibundi) [inline test/serialization.js]: We should add explicit `source` to check against after serialization as a serialize\parse loop may introduce subtle bugs and not check all cases that may be somehow handled during parsing. (just to clarify, checks with serialize/parse loop should remain too)

File: test/serialization.js [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+    config
+  );
+
+  const domains = schema.schemas[0];
+  const stringifyDomains = serializeSchema(domains);
+
+  const category = schema.categories.get('Test');
+  const { type, module, name, source } = serializeSchema(category, {
+    exclude: ({ ins }) =>
+      ins && ['Unique'].includes(metaschema.extractDecorator(ins)),
```
Comment (lundibundi) [inline test/serialization.js]:
```suggestion
      ins && metaschema.extractDecorator(ins) === 'Unique',
```

Comment (lundibundi) [reply test/serialization.js]: Also, there doesn't seem to be `Unique()` fields in serialized schemas.

File: lib/serialization.js [hunk 7]
```diff
@@ -14,16 +14,21 @@ const buildPath = node => {
   return buildPath(node.parentPath);
 };
 
-const getInstance = (path, ins) =>
-  ins && path.length ? getInstance(path, ins[path.shift()]) : ins;
+const getInstance = (path, ins) => {
+  for (let i = 0; ins && i < path.length; i++) {
+    ins = ins[path[i]];
+  }
+
```
Comment (lundibundi) [inline lib/serialization.js]: Nit: unnecessary newline?

File: test/serialization.js [hunk 8]
```diff
@@ -8,7 +8,55 @@ const metaschema = require('..');
...
+    type: 'object',
+    class: 'Uint32Array',
+    length: 10
+  },
+  AnyTypeDomain: {
+    type: 'any'
+  }
+});`;
+
+const categorySource = `({
```
Comment (lundibundi) [inline test/serialization.js]: Could you name these sources according to their respective category/domains?

File: test/serialization.js [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+  );
+
+  stringifyCategory.definition = metaschema.processSchema(
+    'TestCategory',
+    stringifyCategory.source,
+    {},
+    runInNewContext
+  );
+
+  metaschema.Metaschema.create([stringifyDomains, stringifyCategory], config);
```
Comment (lundibundi) [inline test/serialization.js]: You do not seem to check for the correctness of the generated schema though, as you did before by comparing the definitions.

File: test/serialization.js [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+
+const metatests = require('metatests');
+const metaschema = require('..');
+
+const { serializeSchema } = require('../lib/serialization');
+const { options, config } = require('../lib/config');
+
+const decorators = require('../lib/decorators');
+
+const testDomainsSource = `({
```
Comment (nechaido) [inline test/serialization.js]: I'd prefer to remove all whitespaces from serialized sources.

**Comment by lundibundi** [comment]
ping @bugagashenkj

File: lib/schema.js [hunk 11]
```diff
@@ -46,6 +46,26 @@ class Metaschema {
...
   }
 
+  // Serialize an instance of schema
+  //   type <string>
+  //   schema <Schema> | <string> schema or path, that can be resolved by
+  //       config.resolve
+  //   options <Object>
+  // Returns: <Error> | <null>
+  serialize(type, schema, options) {
+    if (!this.config.processors[type]) {
```
Comment (lundibundi) [inline lib/schema.js]:
```suggestion
    const processor = this.config.processors[type];
    if (!processor) {
```
And reuse it later?

File: lib/schema.js [hunk 12]
```diff
@@ -46,6 +46,26 @@ class Metaschema {
...
+      throw new TypeError(`No processors defined for type: '${type}'`);
+    }
+    if (typeof schema === 'string') {
+      schema = this.config.resolve(this, type, schema);
+    }
+    const serializer = this.config.processors[type].serializeSchema;
+    if (!serializer) {
```
Comment (lundibundi) [inline lib/schema.js]: I'd prefer to first check if serializer exists and only then try to resolve the schema by path (`this.config.resolve(this, type, schema);`).

File: lib/schema.js [hunk 13]
```diff
@@ -46,6 +46,26 @@ class Metaschema {
     return errors.length === 0 ? null : new MetaschemaError(errors);
   }
 
+  // Serialize an instance of schema
+  //   type <string>
+  //   schema <Schema> | <string> schema or path, that can be resolved by
+  //       config.resolve
+  //   options <Object>
+  // Returns: <Error> | <null>
```
Comment (lundibundi) [inline lib/schema.js]: This doesn't look correct.

File: lib/schema.js [hunk 14]
```diff
@@ -46,6 +46,27 @@ class Metaschema {
     return errors.length === 0 ? null : new MetaschemaError(errors);
   }
 
+  // Serialize an instance of schema
+  //   type <string>
+  //   schema <Schema> | <string> schema or path, that can be resolved by
+  //       config.resolve
+  //   options <Object>
+  // Returns: <string>
```
Comment (lundibundi) [inline lib/schema.js]:
```suggestion
  // Returns: <string> serialized schema
```

**Pull request merged**